### PR TITLE
[mlir][memref] Fix segfault in SROA.

### LINKED ISSUE #70902: [mlir] --sroa causes Segmentation fault
git version: 604eff60abfce09f956c3b0b1414f8d0d04b5d47

system: ```Ubuntu 20.04.6 LTS (Focal Fossa)```

reproduced with: ```mlir-opt --sroa a.mlir```

a.mlir:
```
func.func private @func1() -> i32 {
  %c13 = arith.constant 13 : index
  %alloca_16 = memref.alloca() : memref<4xi32>
  %114 = memref.load %alloca_16[%c13] : memref<4xi32>
  return %114 : i32
}
```

trace:
```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /data/bin/llvm-project/build/bin/mlir-opt --sroa a.mlir
 #0 0x000055cf680c381f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/data/bin/llvm-project/build/bin/mlir-opt+0x15d981f)
 #1 0x000055cf680c08b4 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f81a50b2420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #3 0x000055cf6ae5c067 mlir::Type::getContext() const (/data/bin/llvm-project/build/bin/mlir-opt+0x4372067)
 #4 0x000055cf6adca778 mlir::MemRefType::get(llvm::ArrayRef<long>, mlir::Type, mlir::MemRefLayoutAttrInterface, mlir::Attribute) (/data/bin/llvm-project/build/bin/mlir-opt+0x42e0778)
 #5 0x000055cf6929d2fd mlir::memref::AllocaOp::destructure(mlir::DestructurableMemorySlot const&, llvm::SmallPtrSetImpl<mlir::Attribute> const&, mlir::RewriterBase&) (/data/bin/llvm-project/build/bin/mlir-opt+0x27b32fd)
 #6 0x000055cf69284b03 mlir::detail::DestructurableAllocationOpInterfaceInterfaceTraits::Model<mlir::memref::AllocaOp>::destructure(mlir::detail::DestructurableAllocationOpInterfaceInterfaceTraits::Concept const*, mlir::Operation*, mlir::DestructurableMemorySlot const&, llvm::SmallPtrSetImpl<mlir::Attribute> const&, mlir::RewriterBase&) (/data/bin/llvm-project/build/bin/mlir-opt+0x279ab03)
 #7 0x000055cf6d959303 mlir::DestructurableAllocationOpInterface::destructure(mlir::DestructurableMemorySlot const&, llvm::SmallPtrSetImpl<mlir::Attribute> const&, mlir::RewriterBase&) (/data/bin/llvm-project/build/bin/mlir-opt+0x6e6f303)
 #8 0x000055cf6ac84ba5 destructureSlot(mlir::DestructurableMemorySlot&, mlir::DestructurableAllocationOpInterface, mlir::RewriterBase&, (anonymous namespace)::MemorySlotDestructuringInfo&, mlir::SROAStatistics const&) SROA.cpp:0:0
 #9 0x000055cf6ac8b3da mlir::tryToDestructureMemorySlots(llvm::ArrayRef<mlir::DestructurableAllocationOpInterface>, mlir::RewriterBase&, mlir::SROAStatistics) (/data/bin/llvm-project/build/bin/mlir-opt+0x41a13da)
#10 0x000055cf6ac8b6e4 mlir::SROAPattern::matchAndRewrite(mlir::DestructurableAllocationOpInterface, mlir::PatternRewriter&) const (/data/bin/llvm-project/build/bin/mlir-opt+0x41a16e4)
#11 0x000055cf6d92af68 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) (/data/bin/llvm-project/build/bin/mlir-opt+0x6e40f68)
#12 0x000055cf6ace29ac (anonymous namespace)::GreedyPatternRewriteDriver::processWorklist() GreedyPatternRewriteDriver.cpp:0:0
#13 0x000055cf6ace6e1b mlir::applyPatternsAndFoldGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) (/data/bin/llvm-project/build/bin/mlir-opt+0x41fce1b)
#14 0x000055cf6ac882f4 (anonymous namespace)::SROA::runOnOperation() SROA.cpp:0:0
#15 0x000055cf6ac2db2e mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/data/bin/llvm-project/build/bin/mlir-opt+0x4143b2e)
#16 0x000055cf6ac2e00a mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/data/bin/llvm-project/build/bin/mlir-opt+0x414400a)
#17 0x000055cf6ac2e834 mlir::PassManager::run(mlir::Operation*) (/data/bin/llvm-project/build/bin/mlir-opt+0x4144834)
#18 0x000055cf6ac1f5bb performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
#19 0x000055cf6ac20a74 processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::MlirOptMainConfig const&, mlir::DialectRegistry&, llvm::ThreadPool*) MlirOptMain.cpp:0:0
#20 0x000055cf6ac20c24 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
#21 0x000055cf6ad1d174 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) (/data/bin/llvm-project/build/bin/mlir-opt+0x4233174)
#22 0x000055cf6ac196d7 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/data/bin/llvm-project/build/bin/mlir-opt+0x412f6d7)
#23 0x000055cf6ac20d71 mlir::MlirOptMain(int, char**, llvm::StringRef, llvm::StringRef, mlir::DialectRegistry&) (/data/bin/llvm-project/build/bin/mlir-opt+0x4136d71)
#24 0x000055cf6ac211d6 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (/data/bin/llvm-project/build/bin/mlir-opt+0x41371d6)
#25 0x000055cf67ffad9b main (/data/bin/llvm-project/build/bin/mlir-opt+0x1510d9b)
#26 0x00007f81a4b7e083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#27 0x000055cf68093f8e _start (/data/bin/llvm-project/build/bin/mlir-opt+0x15a9f8e)
Segmentation fault (core dumped)
```

### PR DESCRIPTION
**Pull request by Moxinilian**
Fixes #70902.

The out of bounds check in the SROA implementation for MemRef was not actually testing anything because it only operated on a store op which does not trigger the logic by itself. It is now checked for real and the underlying bug is fixed.

I checked the LLVM implementation just in case but this should not happen as out-of-bound checks happen in GEP's verifier there.

### CODE REVIEW

File: mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp [hunk 1]
```diff
@@ -187,14 +187,28 @@ DeletionKind memref::LoadOp::removeBlockingUses(
...
+/// Returns the index of a memref in attribute form, given its indices. Returns
+/// a null pointer if whether the indices form a valid index for the provided
+/// MemRefType cannot be computed.
 static Attribute getAttributeIndexFromIndexOperands(MLIRContext *ctx,
-                                                    ValueRange indices) {
+                                                    ValueRange indices,
+                                                    MemRefType memrefType) {
+  ArrayRef<int64_t> shape = memrefType.getShape();
+  if (indices.size() != memrefType.getShape().size())
+    return {};
```
Comment (joker-eph) [inline mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp:198]: Isn't this covered by the verifier?

File: mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp [hunk 2]
```diff
@@ -187,14 +187,28 @@ DeletionKind memref::LoadOp::removeBlockingUses(
...
-/// Returns the index of a memref in attribute form, given its indices.
+/// Returns the index of a memref in attribute form, given its indices. Returns
+/// a null pointer if whether the indices form a valid index for the provided
+/// MemRefType cannot be computed.
 static Attribute getAttributeIndexFromIndexOperands(MLIRContext *ctx,
-                                                    ValueRange indices) {
+                                                    ValueRange indices,
+                                                    MemRefType memrefType) {
+  ArrayRef<int64_t> shape = memrefType.getShape();
+  if (indices.size() != memrefType.getShape().size())
```
Comment (joker-eph) [inline mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp:197]:
```suggestion
  if (indices.size() != shape.size())
```

File: mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp [hunk 3]
```diff
@@ -187,14 +187,28 @@ DeletionKind memref::LoadOp::removeBlockingUses(
...
+    return {};
   SmallVector<Attribute> index;
-  for (Value coord : indices) {
+  for (size_t i = 0; i < shape.size(); i++) {
+    Value coord = indices[i];
+    int64_t dimSize = shape[i];
```
Comment (joker-eph) [inline mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp:202]:
```suggestion
  for (auto [coord, dimSize] : llvm::zip(indices, shape)) {
```

File: mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp [hunk 4]
```diff
@@ -187,14 +187,28 @@ DeletionKind memref::LoadOp::removeBlockingUses(
...
     IntegerAttr coordAttr;
     if (!matchPattern(coord, m_Constant<IntegerAttr>(&coordAttr)))
       return {};
+    if (!coordAttr.getType().isIndex())
+      return {};
```
Comment (joker-eph) [inline mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp:207]: Why is index an issue?

File: mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp (hunk 1 above)
Comment (Moxinilian) [reply mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp:198]: Yes, it's just a matter of whether or not we assume `getAttributeIndexFromIndexOperands` takes operands from LoadOp and StoreOp. This is probably a safe assumption, if documented.

File: mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp (hunk 4 above)
Comment (Moxinilian) [reply mlir/lib/Dialect/MemRef/IR/MemRefMemorySlot.cpp:207]: Not being an index is an issue. But that is also covered by the verifier.

**Review by joker-eph — APPROVED** [review]

**Pull request merged**